lib/postgrex/simple_connection.ex

```diff
@@ -307,41 +346,51 @@ defmodule Postgrex.SimpleConnection do
       {:ok, protocol} ->
         state = %{state | protocol: protocol}
 
-        with {:noreply, state, _} <- maybe_handle(mod, :handle_connect, [mod_state], state) do
-          {:ok, state}
+        with {:keep_state, state, _} <- maybe_handle(mod, :handle_connect, [mod_state], state) do
+          {:keep_state, state}
         end
 
       {:error, reason} ->
         if state.auto_reconnect do
-          {:backoff, state.reconnect_backoff, state}
+          {:keep_state, state, {{:timeout, :backoff}, state.reconnect_backoff, nil}}
         else
           {:stop, reason, state}
         end
     end
   end
 
-  @doc false
-  def handle_call(msg, from, %{state: {mod, mod_state}} = state) do
-    handle(mod, :handle_call, [msg, from, mod_state], from, state)
+  def handle_event({:timeout, :backoff}, nil, @state, state) do
+    {:keep_state, state, {:next_event, :internal, {:connect, :reconnect}}}
   end
 
-  @doc false
-  def handle_info(:timeout, %{protocol: protocol} = state) do
+  def handle_event({:call, from}, {msg, caller_pid}, @state, %{state: {mod, mod_state}} = state) do
+    # We have to do a hack here to carry the actual caller PID over to the handle_call/3
+    # callback, because gen_statem uses a proxy process to do calls with timeout != :infinity.
+    # This results in the caller PID not being the same as the PID in the "from" tuple,
+    # so things like Postgrex.Notifications cannot use that "from"'s PID to register
+    # notification handlers. This approach is paired with reconstructing the proper
+    # "from" tuple in the reply/2 function in this module.
+    {from_pid, from_ref} = from
+    callback_from = {from_pid, {from_ref, caller_pid}}
```

Comment:
Generally speaking, the second element of `from` must be opaque and I think we are going to be in trouble if we ask users to look into it. I think we should rather tell users of this library to not rely on `from` and instead explicitly pass the caller too.

Reply:
In other words, I don't think the `from` should be passed on the caller here. Rather the `.listen` function in `notifications` should explicitly pass the `pid`.

Reply:
I sent a similar PR to Oban which may also be affected by this: https://github.com/sorentwo/oban/pull/843

Reply:
@josevalim yes, the second element of from _is_ opaque here outside of `SimpleConnection`. This is to be backwards compatible for folks using the `SimpleConnection` behaviour, so that they're unlikely to have to change their callbacks.

Reply:
That's the point. Your fix has `handle_call` looking at the second element of `from`, but there was no guarantee that would be of any shape whatsoever _before_ this PR.

Reply:
Perhaps you meant to write this:

```suggestion
    # We have to do a hack here to carry the actual caller PID over to the handle_call/3
    # callback, because gen_statem uses a proxy process to do calls with timeout != :infinity.
    # This results in the caller PID not being the same as the PID in the "from" tuple,
    # so things like Postgrex.Notifications cannot use that "from"'s PID to register
    # notification handlers. This approach is paired with reconstructing the proper
    # "from" tuple in the reply/2 function in this module.
    {from_pid, from_ref} = from
    callback_from = {caller_pid, from}
```

That would be a bit more okayish (and backwards compatible).

Reply:
Yeah jeez, that's exactly what I wanted to do. I mixed things up 🙈 Thanks for catching that!

Reply:
(and fixed)

Reply:
Will this change after https://github.com/erlang/otp/pull/7081 ?